### ISSUE TITLE
Hitting "save" with an expired session causes loss of entered data

### ISSUE THREAD

**Issue by alexgleason** [wagtail/wagtail#2199]
If you add/edit a page but your session expires before you can save, clicking save will redirect you to the login screen then upon logging in your entered data won't be there.

**Comment by timheap** [comment]
I can see three solutions to this:

* Make the session login time really, really long. Pro: the user will never be logged out unexpectedly. Con: The user will never be logged out.
* On the edit screen, hit up the backend occasionally on a noop page that does nothing but refresh the session. This will keep the session alive as long as there is an internet connection, but will revert to the old potentially timing out behaviour if the user suspends their computer, or loses internet connection for a period.
* Just before the form is submitted, hit up a noop page to check if the user is logged in. If they are, submit the form as usual. If they are not, display a pop-up / modal login box. Once the user is logged in again, submit as normal.

**Comment by alexgleason** [comment]
I agree with option 3, seems the most foolproof. The first on doesn't cover a user who manually deleted their cookies. The second one doesn't cover a user who shut their laptop screen and opened it again hours later.

**Comment by akx** [comment]
Persisting the form data to the client's `sessionStorage` or `localStorage` (as a form of local draft storage, so to speak) would also be an option.

**Comment by JoshBarr** [comment]
^ this is a good option! I know it's an inconvenience for admins. Crafting a page and accidentally closing the tab, etc. These things happen!

**Comment by alexgleason** [comment]
I like that idea too. :smile: 

Also just want to document here: the blogging platform Medium sends a JSON delta of every change made by the user to the server. It's kind of brilliant how it's structured. Here's an example of a delta where I had the text "Hello world!" then bolded the word "world":

```json
{
  "id": "9429f4e04bde",
  "deltas": [{
    "type": 3,
    "index": 1,
    "paragraph": {
      "name": "2bf8",
      "type": 1,
      "text": "Hello world!",
      "markups": [{
        "type": 1,
        "start": 6,
        "end": 11
      }]
    },
    "verifySameName": true
  }],
  "baseRev": 3
}
```

(I love how it's retained as plain text, and the inline markup is a totally separate layer of data.)

IMO this is the theoretical ideal, but using localStorage is more realistic for now.

**Comment by davecranwell** [comment]
The team here have talked at some length before about auto-saving, rather than having a save button at all. I don't think we're any closer to agreeing to do it, but it would solve this problem.

**Comment by JoshBarr** [comment]
Storing rich text (and all operations) as JSON deltas with an attribute schema like Medium.com would make absolute sense. 

I've implemented something similar for klim.co.nz/retail-fonts/national/

**Comment by soar** [comment]
I just lost huge text in same situation... "Preview" -> Login Screen with user/password prompt -> Empty document page.

**Comment by thibaudcolas** [comment]
With [concurrent editing notifications](https://github.com/wagtail/roadmap/issues/41), we now have a reliable way to detect the user’s session has expired. This is currently available for pages and snippets. If the user’s session has expired, the editing sessions’ Stimulus controllers could be used to catch the 403 response on the "ping" endpoint and display a modal for the user.

Here’s the corresponding Stimulus code: [editing_sessions/module.html#L22-L35](https://github.com/wagtail/wagtail/blob/df08f999453dc02a674bec93937a38cacc4e4d79/wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html#L22-L35).

Something like: `w-swap:error->w-session#handlePingError`, with `handlePingError` deciding what dialog to display.

We have a design for the existing sign-in form within a dialog, though I expect that considering how customisable this is it’d probably be a good first step to display a much simpler "there was an error" modal, with a button for the user to go to the login page in a separate browser tab.

![Sign in form in modal](https://github.com/user-attachments/assets/52f01fb3-a4b3-4b86-9b0d-e8fa2d70076f)

**Comment by thibaudcolas** [comment]
Noting gov.uk research on this concept: [Timeout warning #104](https://github.com/alphagov/govuk-design-system-backlog/issues/104).

**Comment by thibaudcolas** [comment]
Going to close this in favour of #995, with a proposed fix to simply add a confirmation modal to prevent the loss of newly-added content, without the complexity of gracefully handling expired sessions / CSRF token.

There’s room for us to handle this better but the complexity feels high for what is ultimately a corner case -- might revisit this depending on user feedback.
